diff --git a/pkg/server/rpc.go b/pkg/server/rpc.go
index 8b68645..3459ec9 100644
--- a/pkg/server/rpc.go
+++ b/pkg/server/rpc.go
@@ -135,6 +135,7 @@ func (h *rpcHandlerImpl[RequestType, ResponseType]) run(s *RPCServer) {
 
 			case ir := <-requests:
 				if ir == nil {
+					requests = nil
 					continue
 				}
 				if o := s.RequestObserver; o != nil {
@@ -152,6 +153,7 @@ func (h *rpcHandlerImpl[RequestType, ResponseType]) run(s *RPCServer) {
 
 			case claim := <-claims:
 				if claim == nil {
+					claims = nil
 					continue
 				}
 				h.mu.RLock()
diff --git a/pkg/server/stream.go b/pkg/server/stream.go
index cd52415..d02fce2 100644
--- a/pkg/server/stream.go
+++ b/pkg/server/stream.go
@@ -105,6 +105,7 @@ func (h *streamHandler[RecvType, SendType]) run(s *RPCServer) {
 
 			case is := <-requests:
 				if is == nil {
+					requests = nil
 					continue
 				}
 				if time.Now().UnixNano() < is.Expiry {
@@ -115,6 +116,7 @@ func (h *streamHandler[RecvType, SendType]) run(s *RPCServer) {
 
 			case claim := <-claims:
 				if claim == nil {
+					claims = nil
 					continue
 				}
 				h.mu.RLock()
