// babylonjs-hook/dist/lib/babylonjs-hook.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var core_1 = require("@babylonjs/core");
var scene_1 = require("./scene");
var engine_1 = require("./engine");
__export(require("./engine"));
__export(require("./scene"));
/**
 * Register a callback for before the scene renders.
 *
 * @param callback called using onBeforeRender functionality of scene
 * @param mask the mask used to filter observers
 * @param insertFirst if true will be inserted at first position, if false (default) will be last position.
 * @param callOnce only call the callback once
 */
exports.useBeforeRender = function (callback, mask, insertFirst, callOnce) {
    var scene = react_1.useContext(scene_1.SceneContext).scene;
    react_1.useEffect(function () {
        if (scene === null) {
            return;
        }
        var unregisterOnFirstCall = callOnce === true;
        var sceneObserver = scene.onBeforeRenderObservable.add(callback, mask, insertFirst, undefined, unregisterOnFirstCall);
        if (unregisterOnFirstCall !== true) {
            return function () {
                scene.onBeforeRenderObservable.remove(sceneObserver);
            };
        }
    });
};
/**
 * Register a callback for after the scene renders.
 *
 * @param callback called using onBeforeRender functionality of scene
 * @param mask the mask used to filter observers
 * @param insertFirst if true will be inserted at first position, if false (default) will be last position.
 * @param callOnce only call the callback once
 */
exports.useAfterRender = function (callback, mask, insertFirst, callOnce) {
    var scene = react_1.useContext(scene_1.SceneContext).scene;
    react_1.useEffect(function () {
        if (scene === null) {
            return;
        }
        var unregisterOnFirstCall = callOnce === true;
        var sceneObserver = scene.onAfterRenderObservable.add(callback, mask, insertFirst, undefined, unregisterOnFirstCall);
        if (unregisterOnFirstCall !== true) {
            return function () {
                scene.onAfterRenderObservable.remove(sceneObserver);
            };
        }
    });
};
/**
 * Handles creating a camera and attaching/disposing.
 * TODO: add new 4.2 parameters: useCtrlForPanning & panningMouseButton
 * @param createCameraFn function that creates and returns a camera
 * @param autoAttach Attach the input controls (default true)
 * @param noPreventDefault Events caught by controls should call prevent default
 */
exports.useCamera = function (createCameraFn, autoAttach, noPreventDefault) {
    if (autoAttach === void 0) { autoAttach = true; }
    if (noPreventDefault === void 0) { noPreventDefault = true; }
    var scene = react_1.useContext(scene_1.SceneContext).scene;
    var cameraRef = react_1.useRef(null);
    react_1.useEffect(function () {
        if (scene === null) {
            console.warn('cannot create camera (scene not ready)');
            return;
        }
        var camera = createCameraFn(scene);
        if (autoAttach === true) {
            var canvas = scene.getEngine().getRenderingCanvas();
            // This attaches the camera to the canvas
            // https://github.com/BabylonJS/Babylon.js/pull/9192 (keep canvas to work with < 4.2 beta-13)
            camera.attachControl(canvas, noPreventDefault);
        }
        cameraRef.current = camera;
        return function () {
            if (autoAttach === true) {
                var canvas = scene.getEngine().getRenderingCanvas();
                camera.detachControl(canvas);
            }
            camera.dispose();
        };
    }, [scene]);
    return cameraRef.current;
};
exports.default = (function (props) {
    var reactCanvas = react_1.useRef(null);
    var antialias = props.antialias, engineOptions = props.engineOptions, adaptToDeviceRatio = props.adaptToDeviceRatio, sceneOptions = props.sceneOptions, onRender = props.onRender, onSceneReady = props.onSceneReady, renderChildrenWhenReady = props.renderChildrenWhenReady, children = props.children, rest = __rest(props, ["antialias", "engineOptions", "adaptToDeviceRatio", "sceneOptions", "onRender", "onSceneReady", "renderChildrenWhenReady", "children"]);
    var _a = react_1.useState({
        scene: null,
        sceneReady: false
    }), sceneContext = _a[0], setSceneContext = _a[1];
    var _b = react_1.useState({
        engine: null,
        canvas: null
    }), engineContext = _b[0], setEngineContext = _b[1];
    react_1.useEffect(function () {
        if (reactCanvas.current) {
            var engine_2 = new core_1.Engine(reactCanvas.current, antialias, engineOptions, adaptToDeviceRatio);
            setEngineContext(function () { return ({
                engine: engine_2,
                canvas: reactCanvas.current
            }); });
            var scene_2 = new core_1.Scene(engine_2, sceneOptions);
            var sceneIsReady_1 = scene_2.isReady();
            if (sceneIsReady_1) {
                props.onSceneReady(scene_2);
            }
            else {
                scene_2.onReadyObservable.addOnce(function (scene) {
                    props.onSceneReady(scene);
                    setSceneContext(function () { return ({
                        canvas: reactCanvas.current,
                        scene: scene,
                        engine: engine_2,
                        sceneReady: true,
                    }); });
                });
            }
            engine_2.runRenderLoop(function () {
                if (scene_2.activeCamera) {
                    if (typeof onRender === 'function') {
                        onRender(scene_2);
                    }
                    scene_2.render();
                }
                else {
                    // @babylonjs/core throws an error if you attempt to render with no active camera.
                    // if we attach as a child React component we have frames with no active camera.
                    console.warn('no active camera..');
                }
            });
            var resize_1 = function () {
                scene_2.getEngine().resize();
            };
            if (window) {
                window.addEventListener('resize', resize_1);
            }
            setSceneContext(function () { return ({
                canvas: reactCanvas.current,
                scene: scene_2,
                engine: engine_2,
                sceneReady: sceneIsReady_1,
            }); });
            return function () {
                scene_2.getEngine().dispose();
                if (window) {
                    window.removeEventListener('resize', resize_1);
                }
            };
        }
    }, [reactCanvas]);
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("canvas", __assign({ ref: reactCanvas }, rest)),
        react_1.default.createElement(engine_1.EngineCanvasContext.Provider, { value: engineContext },
            react_1.default.createElement(scene_1.SceneContext.Provider, { value: sceneContext }, (renderChildrenWhenReady !== true || (renderChildrenWhenReady === true && sceneContext.sceneReady)) &&
                children))));
});
