// babylonjs-hook/dist/lib/engine.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
exports.EngineCanvasContext = react_1.createContext({
    engine: null,
    canvas: null
});
function withEngineCanvasContext(Component) {
    return function BoundComponent(props) {
        return (react_1.default.createElement(exports.EngineCanvasContext.Consumer, null, function (ctx) { return react_1.default.createElement(Component, __assign({}, props, { engineCanvasContext: ctx })); }));
    };
}
exports.withEngineCanvasContext = withEngineCanvasContext;
/**
 * Get the engine from the context.
 */
exports.useEngine = function () { return react_1.useContext(exports.EngineCanvasContext).engine; };
/**
 * Get the canvas DOM element from the context.
 */
exports.useCanvas = function () { return react_1.useContext(exports.EngineCanvasContext).canvas; };
